cpp/dolfin/function/Function.cpp

```diff
@@ -275,6 +275,72 @@ void Function::eval(
   }
 }
 //-----------------------------------------------------------------------------
+void Function::eval_reference(
+    const Eigen::Ref<
+        const Eigen::Array<double, Eigen::Dynamic, 3, Eigen::RowMajor>>& X,
+    Eigen::Ref<Eigen::Array<PetscScalar, Eigen::Dynamic, Eigen::Dynamic,
+                            Eigen::RowMajor>>
+        u) const
+{
+  // if (X.rows() != u.rows())
+  // {
+  //   throw std::runtime_error("Length of array for Function values must be the
+  //   "
+  //                            "same as the number of points.");
+  // }
+
+  // Get mesh
+  assert(_function_space);
+  assert(_function_space->mesh());
+  const mesh::Mesh& mesh = *_function_space->mesh();
+  const int gdim = mesh.geometry().dim();
+  const int tdim = mesh.topology().dim();
+
+  // Get element
+  assert(_function_space->element());
+  const fem::FiniteElement& element = *_function_space->element();
+  const int reference_value_size = element.reference_value_size();
+  const int value_size = element.value_size();
+  const int space_dimension = element.space_dimension();
+
+  // Compute basis on reference element
+
+  // NOTE: This step needs to be generalised for elements other than
+  //       non-manifold Lagrange elements
```

Comment:
You mean "non-affine"?

Reply:
No, I meant manifolds where `tdim != `gdim` are not supported. I've added a check that throws an exception if `tdim != `gdim`.